zebra/zebra_nb_state.c

```diff
@@ -147,218 +164,367 @@ lib_interface_zebra_state_mcast_group_get_elem(struct nb_cb_get_elem_args *args)
+
+	zvrf = zebra_vrf_lookup_by_id(vrf->vrf_id);
+
+	if (args->list_entry == NULL) {
+		afi = AFI_IP;
+		safi = SAFI_UNICAST;
+
+		zrt = zebra_router_find_zrt(zvrf, zvrf->table_id, afi, safi);
+		if (zrt == NULL)
+			return NULL;
+	} else {
+		zrt = RB_NEXT(zebra_router_table_head, zrt);
+		/* vrf_id/ns_id do not match, only walk for the given VRF */
+		while (zrt && zrt->ns_id != zvrf->zns->ns_id)
+			zrt = RB_NEXT(zebra_router_table_head, zrt);
+	}
+
+	return zrt;
 }
 
-int lib_vrf_ribs_rib_get_keys(struct nb_cb_get_keys_args *args)
+int lib_vrf_zebra_ribs_rib_get_keys(struct nb_cb_get_keys_args *args)
 {
-	/* TODO: implement me. */
+	const struct zebra_router_table *zrt = args->list_entry;
+
+	args->keys->num = 2;
+
+	snprintfrr(args->keys->key[0], sizeof(args->keys->key[0]), "%s:%s",
+		   "frr-zebra",
+		   zebra_extract_afi_safi_key(zrt->afi, zrt->safi));
+	snprintfrr(args->keys->key[1], sizeof(args->keys->key[1]), "%" PRIu32,
+		   zrt->tableid);
+
 	return NB_OK;
 }
 
-const void *lib_vrf_ribs_rib_lookup_entry(struct nb_cb_lookup_entry_args *args)
+const void *
+lib_vrf_zebra_ribs_rib_lookup_entry(struct nb_cb_lookup_entry_args *args)
 {
-	/* TODO: implement me. */
-	return NULL;
+	struct vrf *vrf = (struct vrf *)args->parent_list_entry;
+	struct zebra_vrf *zvrf;
+	afi_t afi = AFI_IP;
+	safi_t safi = SAFI_UNICAST;
+
+	zvrf = zebra_vrf_lookup_by_id(vrf->vrf_id);
+
+	return zebra_router_find_zrt(zvrf, zvrf->table_id, afi, safi);
 }
 
 /*
- * XPath: /frr-vrf:lib/vrf/frr-zebra:ribs/rib/route
+ * XPath: /frr-vrf:lib/vrf/frr-zebra:zebra/ribs/rib/route
  */
-const void *lib_vrf_ribs_rib_route_get_next(struct nb_cb_get_next_args *args)
+const void *
+lib_vrf_zebra_ribs_rib_route_get_next(struct nb_cb_get_next_args *args)
 {
-	/* TODO: implement me. */
-	return NULL;
+	const struct zebra_router_table *zrt = args->parent_list_entry;
+	const struct route_node *rn = args->list_entry;
+
+	if (args->list_entry == NULL)
+		rn = route_top(zrt->table);
+	else
+		rn = srcdest_route_next((struct route_node *)rn);
+
+	return rn;
 }
 
-int lib_vrf_ribs_rib_route_get_keys(struct nb_cb_get_keys_args *args)
+int lib_vrf_zebra_ribs_rib_route_get_keys(struct nb_cb_get_keys_args *args)
 {
-	/* TODO: implement me. */
+	const struct route_node *rn = args->list_entry;
+
+	args->keys->num = 1;
+	prefix2str(&rn->p, args->keys->key[0], sizeof(args->keys->key[0]));
+
 	return NB_OK;
 }
 
 const void *
-lib_vrf_ribs_rib_route_lookup_entry(struct nb_cb_lookup_entry_args *args)
+lib_vrf_zebra_ribs_rib_route_lookup_entry(struct nb_cb_lookup_entry_args *args)
 {
-	/* TODO: implement me. */
-	return NULL;
+	const struct zebra_router_table *zrt = args->parent_list_entry;
+	struct prefix p;
+	struct route_node *rn;
+
+	yang_str2prefix(args->keys->key[0], &p);
+
+	rn = route_node_lookup(zrt->table, &p);
+	if (!rn) {
+		if (IS_ZEBRA_DEBUG_EVENT) {
+			char buf[PREFIX_STRLEN];
+			zlog_debug("prefix %s is not present in route table",
+				   prefix2str(&p, buf, sizeof(buf)));
```

Comment:
the NB state callbacks shouldn't log anything as `debug northbound calbacks state` already provides all information one might need.

Reply:
removed it as per your comment, but this log was meant to provide clear message about interested prefix is not present in route table.